sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala

```diff
@@ -1629,16 +1629,23 @@ class DataSourceV2SQLSuite
     }
   }
 
-  test("SPARK-33435: REFRESH TABLE should invalidate all caches referencing the table") {
+  test("SPARK-33435, SPARK-34099: REFRESH TABLE should refresh all caches referencing the table") {
     val tblName = "testcat.ns.t"
     withTable(tblName) {
       withTempView("t") {
         sql(s"CREATE TABLE $tblName (id bigint) USING foo")
+        sql(s"INSERT INTO $tblName SELECT 0")
         sql(s"CACHE TABLE t AS SELECT id FROM $tblName")
+        checkAnswer(spark.table(tblName), Row(0))
+        checkAnswer(spark.table("t"), Row(0))
+
+        sql(s"INSERT INTO $tblName SELECT 1")
 
         assert(spark.sharedState.cacheManager.lookupCachedData(spark.table("t")).isDefined)
         sql(s"REFRESH TABLE $tblName")
-        assert(spark.sharedState.cacheManager.lookupCachedData(spark.table("t")).isEmpty)
+        assert(spark.sharedState.cacheManager.lookupCachedData(spark.table("t")).isDefined)
```

Comment:
This PR can be considered as a bug fix. Before this PR, `REFRESH TABLE` uncaches all dependencies from the cache.

Reply:
Hmm... curious why is it a bug fix, `REFRESH TABLE` still recaches the table itself right? and uncaches all dependencies is just conservative but not incorrect?

Reply:
> REFRESH TABLE still recaches the table itself right?

Right.

> and uncaches all dependencies is just conservative but not incorrect?

Correct, in term of correctness, this is not a bug.

Reply:
What's the behavior of 3.0? do we recache or simply uncache?

Reply:
ah it's a new behavior for v2 table in 3.1. What's the behavior for v1 table?

Reply:
Uncaching (as opposed to recaching) dependencies has been the behavior for v1 since 2.0.0 via [SPARK-15367](https://issues.apache.org/jira/browse/SPARK-15367). We followed the same behavior when adding the logic to v2 in #30359.

Reply:
@MaxGekk what's the motivation to change the behavior only for v2 tables?

Reply:
The number of cases that need to handle in `DataSourceV2Strategy` - just one `DataSourceV2Relation`.

The main motivation for such changes is to "fix" `CatalogImpl.refreshTable()` too because it brought significant overhead even for uncached tables since we started to use it in fixing bugs in v1 DDL commands. Just in case, `CatalogImpl.refreshTable()` can be used for v1 as well as for v2 tables (available as public API to users).

 > the behavior only for v2 tables

hmm, you probably asking about v1 DDL commands like `ALTER TABLE .. ADD/DROP PARTITION` (touched in this PR) but recaching of v1 tables didn't work at all a couple weeks ago for some commands and v1 tables. I don't think that comparison of changes here and in `CatalogImpl.refreshTable()` is fair. If we look at the 2.4 branch, it still has troubles.

Reply:
To be clear: does this PR have behavior change?

Reply:
Should not. Results of queries are the same. Since it keeps dependences cached, second actions should be faster.